### ISSUE TITLE
OpenJML CharSequence specification or handling of null in ESC is incorrect

### ISSUE THREAD

**Issue by leavens** [OpenJML/OpenJML#768]
When I try to check the attached specification of SpecifyingLoopsExample1.java with ESC I get the following output
```
-*- mode: compilation; default-directory: "/mnt/c/cygwin64/home/leavens/research/specification-examples/" -*-
Compilation started at Thu Mar 31 13:33:33

esc SpecifyingLoopsExample1.java 
SpecifyingLoopsExample1.java:28: verify: The prover cannot establish an assertion (InvariantEntr```ance: /usr/local/openjml/specs/java/lang/CharSequence.jml:30:) in method main: (Caller: SpecifyingLoopsExample1.main(java.lang.String[]), Callee: SpecifyingLoopsExample1.reverseWord(java.lang.String))
            String tres = inst.reverseWord(teststr);
                                           ^
/usr/local/openjml/specs/java/lang/CharSequence.jml:30: verify: Associated declaration: SpecifyingLoopsExample1.java:28:
	//-RAC@ public invariant charArray != null; // FIXME - why this when there is a non_null declaration
	               ^
2 verification failures

Compilation exited abnormally with code 6 at Thu Mar 31 13:33:57
```

So I have two questions:
1. Why is argv not assumed to have non-null elements?  Also, even when I specify that, as in the attached, it still seems to have an issue.
2. Why is CharSequence having an invariant failure, when I was not using CharSequence and did not write the specification for it?

See the attached for the source file.  This is checked using OpenJML 0.17.0-alpha-12.

[SpecifyingLoopsExample1.zip](https://github.com/OpenJML/OpenJML/files/8392242/SpecifyingLoopsExample1.zip)

**Comment by leavens** [comment]
I was using Ubuntu for this execution of ESC...

**Comment by davidcok** [comment]
Hi Gary,

For 'main' to pass you also need the precondition
//@ requires \forall \bigint i; 0 <= i < argv.length; \invariant_for(argv[i]);
as well as the precondition you gave.

1) Array elements are not currently assumed non-null for the reasons being discussed  in https://github.com/JavaModelingLanguage/RefMan/issues/53
namely, I find it an unworkable default when creating arrays. Plausibly we could adopt the non-null element default for formal parameters.

2) The precondition I give above says that all the array elements must satisfy their static invariants. That assumption was missed.

String is a CharSequence. The charArray is a model field for CharSequence that models the contents of the CharSequence/String. I run into this same error message when doing verification with Strings -- and intend to fix it, but I want to change the charArray model field from a char[] to a \string (which is a \seq of char), which someday OpenJML will map onto SMT string. A \string (seq) won't have this null problem with missing invariants.

**Comment by leavens** [comment]
Hi David,

Thanks, I didn't realize that one had to specify that the invariant holds for each array element also...


        Regards,

        Gary T. Leavens
        329 L3Harris Center
        Computer Science, University of Central Florida
        4000 Central Florida Blvd., Orlando, FL 32816-2362 USA
        http://www.cs.ucf.edu/~leavens phone: +1-407-823-4758
        ***@***.******@***.***>

**Comment by davidcok** [comment]
Fixed the crash for release21-0.6. Also implemented an implicit pre-condition for arrays that all the array elements satisfy their type invariant.  At present just for 1-D arrays.
